Add tutorial notebook for public released object catalog

yymao opened the pull request:
This PR adds a tutorial notebook for public released object catalog. For this notebook to run, one needs to get GCR v0.9.2 and also gcr-catalogs master + LSSTDESC/gcr-catalogs#523 + LSSTDESC/gcr-catalogs#524. 

The notebook can use some polish and maybe a few more science examples. I am opening this PR now so that people can make suggestions (esp. since Slack is down right now). 

I will open another PR for the tutorial notebook for public released truth-match catalog.

yymao committed:
add object notebook

yymao committed:
Add more explanation

yymao committed:
bug fix

yymao committed:
tract_filter, sample_filter moved to helpers

yymao committed:
add .gitignore

yymao committed:
Use preset catalog config

yymao commented:
Just a note for potential reviewers: you should have access to `~yymao/desc` on NERSC if you are in the DESC, so by running the first cell,

```python
import sys
sys.path.insert(0, "/global/homes/y/yymao/desc/generic-catalog-reader")
sys.path.insert(0, "/global/homes/y/yymao/desc/gcr-catalogs/")
```

you'll get all the needed dependencies, and you don't need to grab GCR and GCRCatalogs PRs on your own. Hopefully this will encourage people to review this notebook. 

BTW, those who review this notebook may also want to review https://github.com/LSSTDESC/gcr-catalogs/pull/523, a PR that introduces some helper functions for the purpose of this notebook.

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "The definitions, units, and types of these fields is documented in the data release note (Table 1).\n",
+    "As explained in the note, the values exposed here are not the native columns produced by the Data Management stack.\n",
+    "Instead, this schema strives to mimic the schema of the LSST Data Products Definition Document [DPDD](http://ls.st/dpdd), \n",
+    "which is an effort made by the Rubin to standardize the format of the official Data Release Products (DRP).\n",
```
the Rubin to standardize... To me it looks like jargon, maybe change to "the Rubin commmunity". Or maybe project, observatory, or scientific collaborations?

yymao replied:
Should be "Rubin Observatory". Fixed.

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+   "source": [
+    "### Accessing data in an interactive session\n",
+    "\n",
+    "The DC2 DR6 is a very large data set; loading the full object table into memory is usually not feasible. To access the data in a memory efficient fashion while maintain reasonable performace, we usually need to do one or a combination of the following measures. \n",
```
I think `while maintain` would be `while maintaining`

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "2. Loading a subset of the sky area that you immediately need, or iterating over sky areas (\"tracts\"; see below)\n",
+    "3. Sampling the table at random if your science allows\n",
+    "\n",
+    "Becuase the physical data files are in Parquet format, loading specific columns (1) is efficient and does not require reading in the full table. \n",
```
Becuase -> because

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "3. Sampling the table at random if your science allows\n",
+    "\n",
+    "Becuase the physical data files are in Parquet format, loading specific columns (1) is efficient and does not require reading in the full table. \n",
+    "The physical data files are also partitioned in \"tracts\", each of which corresponds to a region of the sky (see Figure 1 of the release note); hence, loading only a subset of \"tracts\" or iterate over tracts (2) will also make the memory footprint much smaller. \n",
```
`iterate` -> `iterating`

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "\n",
+    "Becuase the physical data files are in Parquet format, loading specific columns (1) is efficient and does not require reading in the full table. \n",
+    "The physical data files are also partitioned in \"tracts\", each of which corresponds to a region of the sky (see Figure 1 of the release note); hence, loading only a subset of \"tracts\" or iterate over tracts (2) will also make the memory footprint much smaller. \n",
+    "Finally, subsampling the table at random can make your analysis calculation faster and use less memory, but it does not save much in term of I/O as all the rows need to be read in before subsampling.\n",
```
I would leave either calculation or analysis and take the other out.

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "\n",
+    "#### If you use GCRCatalogs to access the data\n",
+    "\n",
+    "You will use `get_quantities` to retrive the data, and there are several convient, high-level features that help you achieve the three measures described above. When calling `get_quantities`, you can\n",
```
`retrive` -> `retrieve`

fjaviersanchez replied:
`convient` -> `convenient`

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "You can read more about how to specify the queries [here](https://github.com/yymao/easyquery#usage), \n",
+    "but we will show you some common examples below.\n",
+    "\n",
+    "Each query should be a string that contains a *simple* python expression, in which the variables are column names. \n",
```
Maybe include a link to an explanation of what a simple python expression is.

fjaviersanchez replied:
https://docs.python.org/3/reference/simple_stmts.html

yymao replied:
Instead I added a reference to [numexpr doc](https://numexpr.readthedocs.io/en/latest/user_guide.html#supported-operators) for supported operators and functions, since that's what GCRQuery actually uses.

fjaviersanchez commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "* Fluxes and magnitudes (PSF and [CModel](https://www.sdss.org/dr12/algorithms/magnitudes/#cmodel))\n",
+    "* Adaptive second moments (using [GalSim's HSM](http://galsim-developers.github.io/GalSim/namespacegalsim_1_1hsm.html))\n",
+    "* Quality flags: e.g, does the source have any interpolated pixels? Has any of the measurement algorithms returned an error?\n",
+    "* Other useful quantities: `blendedness`, measure of how flux is affected by neighbors: (1 - flux.child/flux.parent) (see 4.9.11 of [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B)); `extendedness`, classifies sources as extended or psf-like."
```
Maybe say that more details on `extendedness` are later in the notebook?

yymao replied:
I've added the reference for `extendedness`, since we also gave the reference to `blendedness` here too.

fjaviersanchez reviewed (changes requested):
Sorry, I missed this yesterday. I think that the notebook looks good. There are a few typos but the plots and code look good and relevant to me. 

As a stylistic suggestion: Maybe I would add a phrase towards the end to say that it's the end of the notebook: "Finally, we show a practical example to illustrate how to access the dataset and perform a specific selection. We have `extendedness`...". In any case, please feel free to disregard this last suggestion.

yymao committed:
fix typos and implement suggestions

yymao commented:
Thanks, @fjaviersanchez. I've fixed the typos and implemented your suggestions. 

I'll also leave this PR open for a bit in case others have further comments. Similarly we need to remove/update the custom paths in the PR when the download portol is set up.

fjaviersanchez reviewed (approved):
Thanks!

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The definitions, units, and types of these fields is documented in the data release note (Table 1).\n",
```
```suggestion
    "The definitions, units, and types of these fields are documented in the data release note (Table 1).\n",
```

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+   "metadata": {},
+   "source": [
+    "The definitions, units, and types of these fields is documented in the data release note (Table 1).\n",
+    "As explained in the note, the values exposed here are not the native columns produced by the Data Management stack.\n",
```
```suggestion
    "As explained in the note, the values exposed here are not the native columns produced by the LSST Science Pipelines.\n",
```

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "* Positions\n",
+    "* Fluxes and magnitudes (PSF and [CModel](https://www.sdss.org/dr12/algorithms/magnitudes/#cmodel))\n",
+    "* Adaptive second moments (using [GalSim's HSM](http://galsim-developers.github.io/GalSim/namespacegalsim_1_1hsm.html))\n",
+    "* Quality flags: e.g, does the source have any interpolated pixels? Has any of the measurement algorithms returned an error?\n",
```
```suggestion
    "* Quality flags: e.g, does the source have any interpolated pixels? Has any measurement algorithm returned an error?\n",
```
or
```suggestion
    "* Quality flags: e.g, does the source have any interpolated pixels? Have any of the measurement algorithms returned an error?\n",
```

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
````diff
@@ -0,0 +1,554 @@
+    "There are three ways to call `tract_filter`\n",
+    "\n",
+    "```python\n",
+    "tract_filter(4225)  # select one tract, 4225\n",
````
At some point it might be useful to mention how the user can get a list of all available tracts?

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For now, we have `extendedness` as a tool for star/galaxy classification. An object is considered extended if the the difference between the `PSF` magnitude and the [`CModel` magnitude](https://www.sdss.org/dr12/algorithms/magnitudes/#cmodel) is beyond certain threshold (0.0164). To know more about this see [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B) section 4.9.10 ...
```
```suggestion
    "For now, we have `extendedness` as a tool for star/galaxy classification. An object is considered extended if the the difference between the `PSF` magnitude and the [`CModel` magnitude](https://www.sdss.org/dr12/algorithms/magnitudes/#cmodel) is above a certain threshold (0.0164). To learn more about this see [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B) section 4.9.10.\n",
```

rmandelb commented on notebooks/object_table_gcrcatalogs_intro.ipynb:
```diff
@@ -0,0 +1,554 @@
+    "* Fluxes and magnitudes (PSF and [CModel](https://www.sdss.org/dr12/algorithms/magnitudes/#cmodel))\n",
+    "* Adaptive second moments (using [GalSim's HSM](http://galsim-developers.github.io/GalSim/namespacegalsim_1_1hsm.html))\n",
+    "* Quality flags: e.g, does the source have any interpolated pixels? Has any of the measurement algorithms returned an error?\n",
+    "* Other useful quantities: `blendedness`, measure of how flux is affected by neighbors: (1 - flux.child/flux.parent) (see 4.9.11 of [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B)); `extendedness`, classifies sources as extended or point source (see 4.9.10 of [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B))."
```
```suggestion
    "* Other useful quantities: `blendedness`, measure of how flux is affected by neighbors: (1 - flux.child/flux.parent) (see section 4.9.11 of [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B)); `extendedness`, classifies sources as extended or point source (see section 4.9.10 of [Bosch et al. 2018](http://adsabs.harvard.edu/abs/2018PASJ...70S...5B))."
```

rmandelb reviewed (commented):
Hi Yao.  Thanks, this is super useful!  I was wondering if it might be worth incorporating a plot showing number counts as a function of magnitude? (I know you have the color plot, but i thought showing that the counts do what you'd expect as a function of magnitude when you select galaxies might be of interest to users - kind of like in https://github.com/LSSTDESC/DC2-analysis/blob/master/contributed/Run_1.2_exploration.ipynb though there we also had truth catalogs, which I realize is not part of this notebook.  At least when I'm playing with a new dataset I consider that a key plot...)

yymao committed:
Update notebooks/object_table_gcrcatalogs_intro.ipynb

Co-authored-by: Rachel Mandelbaum <rmandelb@andrew.cmu.edu>

yymao committed:
Update notebooks/object_table_gcrcatalogs_intro.ipynb

Co-authored-by: Rachel Mandelbaum <rmandelb@andrew.cmu.edu>

yymao committed:
Update notebooks/object_table_gcrcatalogs_intro.ipynb

Co-authored-by: Rachel Mandelbaum <rmandelb@andrew.cmu.edu>

yymao committed:
Update notebooks/object_table_gcrcatalogs_intro.ipynb

Co-authored-by: Rachel Mandelbaum <rmandelb@andrew.cmu.edu>

yymao committed:
Update notebooks/object_table_gcrcatalogs_intro.ipynb

Co-authored-by: Rachel Mandelbaum <rmandelb@andrew.cmu.edu>

yymao committed:
address review comments

yymao commented:
Thanks @rmandelb --- in addition to the typo/wording fixes which I just accepted your suggestions, I've also added how to find out available tracts and the example of making a galaxy number density plot (mostly stole from your notebook).

rmandelb commented:
Thanks!  I think this notebook will be super useful.

yymao committed:
change mag bin range

yymao committed:
Update root_dir path
